### PR TITLE
Reimplemented pathoscope.run in rust

### PR DESCRIPTION
**Pull request by SnickerDoodlePop**
Implimented many of the pathoscope.py functions in rust with the goal of improving runtime performance; added a binding for expectation_maximization::run (.rs) using pyo3 and maturin.

Output data of virtool_expectation_maximization::run (.whl) is not identical to the previous version of pathoscope.run. Data is minutely different in the last 2-3 floating point digits; the output matrix has records in different indices, but fields still correspond to the correct record.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,946 @@
...
+
+///Wrapper for the buildMatrix function and adjacent code
+mod BuildMatrix
+{
+    use std::collections::HashMap;
+    use std::fs::File;
+    use std::io::BufReader;
+    use super::ParseSam::*;
+
+    ///produces the SAM matrix given a path to a .SAM file
```
Comment (igboyes) [inline src/lib.rs:11]:
```suggestion
```

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,946 @@
+#![allow(nonstandard_style)]
+
+///Wrapper for the buildMatrix function and adjacent code
```
Comment (igboyes) [inline src/lib.rs:3]:
```suggestion
```

**Review by igboyes — CHANGES_REQUESTED** [review]
Thanks. I am excited to see this perform in production.

# Housekeeping

* Clean-up various variants of `rewrite.sam`. Make sure there are no unused test files.
* Eschew non-standard formatting. Use automated formatting provided by `rustfmt`.
* Fix inconsistencies in whitespace. `rustfmt` will deal with this for you.
* Is there a benefit to declaring `mod` with only a single function in it?
* Is there any unused Python code remaining in `pathoscope.py`? Please remove if so.

# Tests
For the most part, I don't think the Rust unit tests are thorough enough to guarantee there has been no regression as a result of future changes.

We should be verifying that given inputs result in exactly the same output. The current assertions only touch part of the output where we should be verifying the entire output. This will provide a short feedback loop to support fearless refactors of these functions in the future.

# Safety
I think it is unlikely that you need to use `unsafe` for this work. If the code needs to start deviating from the patterns from the old Python module in order for it to have safety guarantees, so be it.

File: tests/test_pathoscope.py [hunk 3]
```diff
@@ -57,10 +57,10 @@ def test_build_matrix(data_regression, datafiles):
...
 
 @pytest.mark.datafiles(SAM_PATH)
-@pytest.mark.parametrize("theta_prior", [0, 1e-5])
-@pytest.mark.parametrize("pi_prior", [0, 1e-5])
-@pytest.mark.parametrize("epsilon", [1e-6, 1e-7, 1e-8])
-@pytest.mark.parametrize("max_iter", [5, 10, 20, 30])
+@pytest.mark.parametrize("theta_prior", [0])
+@pytest.mark.parametrize("pi_prior", [0])
+@pytest.mark.parametrize("epsilon", [1e-6])
+@pytest.mark.parametrize("max_iter", [5])
```
[the remark this answers is not in the document]
Comment (SnickerDoodlePop) [reply tests/test_pathoscope.py:63]: oops, this wasn't supposed to be committed

**Review by igboyes — APPROVED** [review]

**Pull request merged**